docs/contributor-guide/how to/test-rules.md

```diff
@@ -0,0 +1,137 @@
+# Test a rule
+
+If you have used the built-in tools to create a new rule (internal or
+external), everything should already set up to use `rule-runner.ts`
+and the `testRule` method.
+
+If not, you just need to:
+
+1. Create a `tests.ts` file in a folder with the name of the rule
+   (e.g.: `src/tests/rules/<rule-id>/tests.ts`)
+1. Have the following template:
+
+```ts
+import { RuleTest } from '../../../helpers/rule-test-type'; // eslint-disable-line no-unused-vars
+import * as ruleRunner from '../../../helpers/rule-runner';
+import { getRuleName } from '../../../../src/lib/utils/rule-helpers';
+
+const ruleName = getRuleName(__dirname);
+
+const tests: Array<RuleTest> = [
+    {
+        name: 'Name of the tests',
+        serverConfig: 'HTML to use',
+        reports: [{
+            message: 'Message the error will have',
+            position: { column: 0, line: 0 } // Where the error will show.
+        }]
+    },
+    { ... }
+];
+
+ruleRunner.testRule(ruleName, tests);
+```
+
+## testRule()
+
+The signature of `ruleRunner.testRule` is:
+
+* `ruleName`, the name of the rule.
+* `tests`, an `Array<RuleTest>`.
+* `ruleConfig`, (optional) to modify the defaults of the rule.
+* `serial`, (optional, defaults to `false`) to run the tests of that
+  rule serially.
+
+`serverConfig` can be of different types depending on particular needs:
+
+* `string` containing the response for `/` (HTML, plain text, etc.).
+* `object` with paths as properties names and their content as values:
+
+<!-- eslint-disable no-unused-vars -->
+
+```js
+const serverConfig = {
+    '/': 'some HTML here',
+    'site.webmanifest': 'other content'
+};
+```
+
+You can even specify the headers and status code for the response for
+a specific path, by using the `headers` and `status` properties:
```

Comment:
Should we also talk about the more complex option which allows conditional response based on the request?

Reply:
Yes, I haven't used that part but if you can point me to a test I can add it.

Reply:
See: https://github.com/sonarwhal/sonarwhal/blob/a2f201e22d84212173b5e4fef82776c7410bdee6/tests/helpers/test-server.ts#L174-L211